### PR TITLE
Day1

### PR DESCRIPTION
**Pull request by DanielBruenker**

### CODE REVIEW

File: src/main/java/org/haffson/adventofcode/utils/FileReaders.java [hunk 1]
```diff
@@ -14,4 +13,18 @@
...
+        InputStream inputStream = FileReaders.class.getResourceAsStream(path);
+        Scanner scanner = new Scanner(inputStream);
+        while (scanner.hasNextLine()) {
+            String line = scanner.nextLine();
+            if(!line.isEmpty()){
+              data.add(line);
+            }
+        }
+        scanner.close();
+        return data.stream().mapToInt(Integer::parseInt).toArray();
```
Comment (MarvinSchoening) [inline src/main/java/org/haffson/adventofcode/utils/FileReaders.java:28]: Macht in diesem Falle keinen großen visuellen Unterschied, aber im Allgemein bevorzugen wir es bei streams die einzelnen aufrufe untereinander zu schreiben, um es übersichtlich zu halten.

```suggestion
        return data.stream()
                   .mapToInt(Integer::parseInt)
                   .toArray();
```

File: src/main/java/org/haffson/adventofcode/days/day01/Day01.java [hunk 2]
```diff
@@ -41,21 +40,94 @@ public int getDay() {
...
 
     /**
      * Primary method for Day 1, Part 1.
-     * Calculates the final frequency as the sum of all frequencies.
+     * Calculates the Report
+     *
+     * @return Returns the product of the multiplication of two numbers that add up to 2020.
+     */
+    private int calculatePart1() {
+        int[] entries = this.findTwoEntriesThatSumToX(this.numbers, 2020);
```
Comment (mfbieber) [inline src/main/java/org/haffson/adventofcode/days/day01/Day01.java]: The `this` keyword is not really needed in Java, but it also does not do any harm.

File: src/main/java/org/haffson/adventofcode/utils/FileReaders.java (hunk 1 above)
Comment (mfbieber) [reply src/main/java/org/haffson/adventofcode/utils/FileReaders.java:28]: In such short stream calls, I would not enforce that suggestion though!

File: src/main/java/org/haffson/adventofcode/days/day01/Day01.java [hunk 3]
```diff
@@ -41,21 +40,94 @@ public int getDay() {
...
+     * This helper function is used to find two list entries that add up to the value x.
      *
-     * @return the final frequency
+     * @param values values
+     * @param x sum of the two list entries
+     * @return Returns the list entries witch add up to x.
      */
-    private int calculateFrequency() {
-        return 0;
+    private int[] findTwoEntriesThatSumToX(int[] values, int x) {
```
Comment (mfbieber) [inline src/main/java/org/haffson/adventofcode/days/day01/Day01.java]: You could turn these methods into package-private methods and test the business logic in unit tests explicitly. There you could also try to think about edge cases and whether your code covers them without running into errors.

File: src/test/java/org/haffson/adventofcode/days/day01/Day01Test.java [hunk 4]
```diff
@@ -4,34 +4,43 @@
...
 
-    @MockBean
+    @Autowired
     private FileReaders fileReaders;
 
-    @Test
+    @Value("${day1.file}")
+    private String filePath;
+
     public void testGetDay() {
```
Comment (mfbieber) [inline src/test/java/org/haffson/adventofcode/days/day01/Day01Test.java:26]: See the comment above about testing the business logic separately 😄 
In addition, you could add tests that verify these simple example inputs from the advent of code problems. The problems will get more complicated and also will compute for longer than a few seconds. If you have these (and many) small tests with small datasets, you can verify that your code works in principle, before running it with the large input files.

File: src/main/java/org/haffson/adventofcode/days/day01/Day01.java [hunk 5]
```diff
@@ -41,21 +40,94 @@ public int getDay() {
...
+     *
+     * @param values values
+     * @return Returns the product of all list entries.
+     */
+    private int multiplyListEntries(int[] values) {
+        int result = 1;
+        for (int value : values) {
+            result = result * value;
+        }
+        return result;
```
Comment (mfbieber) [inline src/main/java/org/haffson/adventofcode/days/day01/Day01.java]: You could simplify this to `return Arrays.stream(values).reduce(1, (a, b) -> a * b);`. In general, there are a lot of useful operations on streams available, you should check them out 😄

**Review by mfbieber — COMMENTED** [review]
Looks good overall! I would just strongly encourage to test the business logic for the puzzle with some edge cases to be sure that this does not break, as you did well with the file reader.

For example, what happens when `findThreeEntriesThatSumToX` returns null in `multiplyListEntries`? Is the result still meaningful? Can you construct cases where the loops break (your code looks fine to me without having tested it, this is just a general question)?

File: src/main/java/org/haffson/adventofcode/utils/FileReaders.java (hunk 1 above)
Comment (DanielBruenker) [reply src/main/java/org/haffson/adventofcode/utils/FileReaders.java:28]: True, it is more readable.

File: src/test/java/org/haffson/adventofcode/utils/FileReadersTest.java [hunk 6]
```diff
@@ -1,18 +1,32 @@
 package org.haffson.adventofcode.utils;
 
+import org.junit.Assert;
```
Comment (mfbieber) [inline src/test/java/org/haffson/adventofcode/utils/FileReadersTest.java:3]: Note here that you are using the assertions from the `org.junit` package. This is perfectly fine, but most of the time we prefer to use AssertJ since it has many more powerful assertions. Maybe you can test them in a subsequent problem you solve.

File: src/test/java/org/haffson/adventofcode/days/day01/Day01Test.java (hunk 4 above)
Comment (DanielBruenker) [reply src/test/java/org/haffson/adventofcode/days/day01/Day01Test.java:26]: You're right. I need to write more tests for the special cases.

File: src/main/java/org/haffson/adventofcode/days/day01/Day01.java (hunk 5 above)
Comment (DanielBruenker) [reply src/main/java/org/haffson/adventofcode/days/day01/Day01.java]: Thanks, `Arrays.stream(values).reduce(1, (a, b) -> a * b); ` is more elegant.

File: src/test/java/org/haffson/adventofcode/utils/FileReadersTest.java (hunk 6 above)
Comment (DanielBruenker) [reply src/test/java/org/haffson/adventofcode/utils/FileReadersTest.java:3]: Okay, then I will switch to  AssertJ.

File: src/main/java/org/haffson/adventofcode/days/day01/Day01.java [hunk 7]
```diff
@@ -4,29 +4,36 @@
...
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
 import org.springframework.stereotype.Component;
 
+import java.util.Arrays;
 import java.util.HashMap;
 
 /**
- * Implementation for <i>Day 1: Chronal Calibration</i>.
+ * Implementation for <i>Day 1: Report Repair/i>.
```
Comment (MatthiasMolitor) [inline src/main/java/org/haffson/adventofcode/days/day01/Day01.java]: Minor issue:  `<` is missing for the closing tag

File: src/main/java/org/haffson/adventofcode/utils/FileReaders.java [hunk 8]
```diff
@@ -14,4 +13,18 @@
...
+        List<String> data = new ArrayList<>();
+        InputStream inputStream = FileReaders.class.getResourceAsStream(path);
+        Scanner scanner = new Scanner(inputStream);
+        while (scanner.hasNextLine()) {
+            String line = scanner.nextLine();
+            if(!line.isEmpty()){
+              data.add(line);
+            }
+        }
+        scanner.close();
```
Comment (MatthiasMolitor) [inline src/main/java/org/haffson/adventofcode/utils/FileReaders.java]: By using try-with-resources statements, you can ensure that resources are **always** released. 
Even in case of exceptions:

```
try (Scanner scanner = new Scanner(inputStream) {
        while (scanner.hasNextLine()) {
               // ...
    }
}
```

try-with-resources  makes manual ``close()`` calls obsolete.
``close()`` will be called automatically, even if an exception is thrown within the ``try`` block.

File: src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 9]
```diff
@@ -0,0 +1,8 @@
+{
+  "properties": [
+    {
+      "name": "day1.file",
+      "type": "java.lang.String",
+      "description": "Path to the puzzle file of day 1."
```
Comment (MatthiasMolitor) [inline src/main/resources/META-INF/additional-spring-configuration-metadata.json:6]: :+1:

File: src/main/java/org/haffson/adventofcode/days/day01/Day01.java [hunk 10]
```diff
@@ -4,29 +4,36 @@
...
+
     @Autowired
-    Day01(FileReaders fileReaders) {
+    Day01(FileReaders fileReaders, @Value("${day1.file}") String filePath) {
         this.problemStatus = new HashMap<>();
-        this.problemStatus.put("1", ProblemStatusEnum.UNSOLVED);
-        this.problemStatus.put("2", ProblemStatusEnum.UNSOLVED);
+        this.problemStatus.put("1", ProblemStatusEnum.SOLVED);
+        this.problemStatus.put("2", ProblemStatusEnum.SOLVED);
+        this.numbers = fileReaders.readFileIntoArrayOfIntegers(filePath);
```
Comment (MatthiasMolitor) [inline src/main/java/org/haffson/adventofcode/days/day01/Day01.java]: You may want to delay reading the file until the numbers are needed.

In this case, the advantage would be an improved startup-time. The disadvantage would be, that you cannot validate the file contents immediately.

Another point: Simple constructors usually simplify testing.

File: src/main/resources/puzzleInputs/Input01.txt [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+1575
+1872
+1261
+1137
+1978
+1537
+1897
+1837
+1753
+1913
```
Comment (MatthiasMolitor) [inline src/main/resources/puzzleInputs/Input01.txt:200]: Versioning the puzzle input with the code is a really good choice here :+1:

**Review by MatthiasMolitor — APPROVED** [review]
Great documentation! Really good solution, touching a lot of Spring Boot stuff :+1:

File: src/main/java/org/haffson/adventofcode/utils/FileReaders.java (hunk 8 above)
Comment (DanielBruenker) [reply src/main/java/org/haffson/adventofcode/utils/FileReaders.java]: Thanks for the tip, I will change the code section.

File: src/main/java/org/haffson/adventofcode/days/day01/Day01.java (hunk 10 above)
Comment (DanielBruenker) [reply src/main/java/org/haffson/adventofcode/days/day01/Day01.java]: The reading of the puzzle values is not optimal in the constructor. I am still looking at how to solve it differently.

File: src/main/java/org/haffson/adventofcode/days/day01/Day01.java [hunk 12]
```diff
@@ -20,6 +21,13 @@
 
     private final int[] numbers;
 
+    Day01() {
+        this.problemStatus = new HashMap<>();
+        this.problemStatus.put("1", ProblemStatusEnum.SOLVED);
+        this.problemStatus.put("2", ProblemStatusEnum.SOLVED);
+        this.numbers = new int[]{};
+    }
+
```
Comment (mfbieber) [inline src/main/java/org/haffson/adventofcode/days/day01/Day01.java]: Are you using this constructor, so that you don't have to use the `FileReaders` in the test? In this case, you could use the other constructor and provide a `FilesReaders` mock. Then, you don't need this extra constructor:

You could try something like this:
````
import org.haffson.adventofcode.days.Days;
import org.haffson.adventofcode.days.day01.Day01;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
public class FileReadersTest {

    @Test
    void test_x(){
        int[] values = new int[]{979, 366, 675};

        FileReaders fileReaders = mock(FileReaders.class);
        when(fileReaders.readFileIntoArrayOfIntegers(anyString()))
                .thenReturn(values);

        int expectedResult = 241861950;

        Days day01 = new Day01(fileReaders, "path");

        int actualResult = day01.multiplyListEntries(values);

        assertThat(actualResult).isEqualTo(expectedResult);
    }

}

````

You will also need the mockito dependency:
````
    testImplementation 'org.junit.jupiter:junit-jupiter:5.6.0'
    testImplementation 'org.mockito:mockito-junit-jupiter:3.3.3'
````

Comment (mfbieber) [reply src/main/java/org/haffson/adventofcode/days/day01/Day01.java]: The problematic thing about this is that you never know what to expect from `numbers`. Does it contain values or not? You end up needing quite some checks in the methods using that variable because you have two constructors that set `numbers` differently. It is better to have just one specific and reliable behavior.

Also, your methods don't use the class variable `numbers` but one that is passed to them via the method parameter. If these methods should be used as util methods, this is fine. In your case, you could just use the class variable. Or you could make the methods static and thus not dependent on an instance of that class. Normally we do this when these are to be used as utility methods.

File: src/test/java/org/haffson/adventofcode/days/day01/Day01Test.java [hunk 13]
```diff
@@ -20,28 +22,108 @@
...
+
+        assertThat(actualResult).isEqualTo(expectedResult);
+    }
+
+    @Test
+    public void testMultiplyListEntriesWithEmptyList() {
+        int[] values = new int[]{};
+        int expectedResult = 0;
+
+        int actualResult = this.day01.multiplyListEntries(values);
```
Comment (mfbieber) [inline src/test/java/org/haffson/adventofcode/days/day01/Day01Test.java]: If you are using your methods like this, and providing the values from the outside (rather than using the class variable `numbers`, see my comment above), you might as well make these methods static and thus not dependent on an instance of the class Day01.

Also note that normally the `this` keyword is not used like this. It is only needed if you have the same variable coming e.g. as a method parameter and want to differentiate if you mean `this` class or (without keyword) the locally scoped variable.
